File: Homework6/Hw6/Parser.fs
module Hw6.Parser

open System
open System.Net
open Hw6.Calculator
open System.Globalization

let parseOperation (operation: string) =
    let oper = operation.Split('=').[1]
    match oper with
    | "Plus" -> Ok CalculatorOperation.Plus
    | "Minus" -> Ok CalculatorOperation.Minus
    | "Multiply" -> Ok CalculatorOperation.Multiply
    | "Divide" -> Ok CalculatorOperation.Divide
    | _ -> Error $"Could not parse value '{WebUtility.UrlDecode(oper)}'"

let parseStringToDouble (arg : string) =
    let argument = arg.Split('=').[1]
    match Double.TryParse(argument, NumberStyles.Float, CultureInfo.InvariantCulture) with
    | (true, value) -> Ok value
    | _ -> Error $"Could not parse value '{argument}'"

File: Homework8/Hw8/Calculator/Parser.cs
using System.Globalization;

namespace Hw8.Calculator;

public class Parser
{
    public static CalculatorArgs ParseCalculatorArgs(string value1, string operation, string value2)
    {
        CalculatorArgs args = new CalculatorArgs();
        args.Value1 = ParseStringToDouble(value1);
        args.Operation = ParseOperation(operation);
        args.Value2 = ParseStringToDouble(value2);

        return args;
    }
    
    private static Operation ParseOperation(string operation)
    {
        return operation.ToLower() switch
        {
            "plus" => Operation.Plus, 
            "minus" => Operation.Minus, 
            "multiply" => Operation.Multiply,
            "divide" => Operation.Divide,
            _ => throw new InvalidOperationException(Messages.InvalidOperationMessage)
        };
    }

    private static double ParseStringToDouble(string val)
    {
        if (!Double.TryParse(val, NumberStyles.Any, CultureInfo.InvariantCulture, out var value))
            throw new ArgumentException(Messages.InvalidNumberMessage);

        return value;
    }
}